File: plantmaintenance/plantmaintenance/report/task_detail_action_flow/task_detail_action_flow.js
// Copyright (c) 2025, LogicalDNA and contributors
// For license information, please see license.txt

frappe.query_reports["Task Detail Action Flow"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1)
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today()
        },
        {
            "fieldname": "task_detail",
            "label": __("Task ID"),
            "fieldtype": "Link",
            "options": "Task Detail",
            "width": 200
        },
        {
            "label": "Status",
            "fieldname": "status",
            "fieldtype": "Select",
            "options":"\nOpen\nIn Progress\nPending Approval\nRejected\nApproved\nCompleted\nCancelled\nOverdue",
            "width": 200,
        },
    ],
    "onload": function(report) {
        frappe.call({
            method: "frappe.client.get_value",
            args: {
                doctype: "User",
                filters: { "name": frappe.session.user },
                fieldname: ["roles"]
            },
            callback: function(response) {
                if (response.message) {
                    let user_roles = frappe.user_roles;
                    
                    let is_maintenance_manager = user_roles.includes("Maintenance Manager");
                    let is_process_manager = user_roles.includes("Process Manager");
    
                    console.log("User Roles:", user_roles);
    
                    let dropdown_html = `
                        <div class="btn-group">
                            <button type="button" class="btn btn-primary dropdown-toggle" data-toggle="dropdown">
                                Workflow Action
                            </button>
                            <div class="dropdown-menu">
                                ${is_maintenance_manager ? `<a class="dropdown-item approve-action" href="#"> Approve</a>
                                <a class="dropdown-item reject-action" href="#"> Reject</a>` : ""}
                                
                                ${is_process_manager ? `<a class="dropdown-item complete-action" href="#"> Complete</a>
                                <a class="dropdown-item cancel-action" href="#"> Cancel</a>` : ""}
                            </div>
                        </div>
                    `;
    
                    if (is_maintenance_manager || is_process_manager) {
                        $(report.page.wrapper).find(".page-actions").prepend(dropdown_html);
                    }
                }
            }
        });
        $(document).on("click", ".approve-action", function() {
            update_task_status("Approved");
        });
    
        $(document).on("click", ".reject-action", function() {
            update_task_status("Rejected");
        });
    
        $(document).on("click", ".complete-action", function() {
            update_task_status("Completed");
        });
    
        $(document).on("click", ".cancel-action", function() {
            update_task_status("Cancelled");
        });
    },    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (column.fieldname === "select_task") {
            let checked = data.select_task ? "checked" : "";
            value = `
                <div class="checkbox-container" data-task="${data.task_detail}" style="cursor: pointer; text-align: center;">
                    <input type="checkbox" class="select-task-checkbox hidden-checkbox" ${checked} style="display: none;">
                    <span class="custom-checkbox" style="display: inline-block; width: 20px; height: 20px; border: 2px solid #000; text-align: center; line-height: 18px; font-size: 16px;">
                        ${checked ? "✔" : ""}
                    </span>
                </div>
            `;
        }
        return value;
    }
};

$(document).on("click", ".checkbox-container", function() {
    let checkbox = $(this).find(".select-task-checkbox");
    let checkboxIcon = $(this).find(".custom-checkbox");

    let isChecked = !checkbox.prop("checked");
    checkbox.prop("checked", isChecked);
    checkboxIcon.html(isChecked ? "✔" : "");

    console.log("Checkbox clicked: ", checkbox.prop("checked"));
});

function update_task_status(status) {
    let selected_tasks = [];

    $(".select-task-checkbox:checked").each(function() {
        let task_id = $(this).closest(".checkbox-container").data("task");
        selected_tasks.push(task_id);
    });

    if (selected_tasks.length === 0) {
        frappe.msgprint(__("Please select at least one task."));
        return;
    }

    frappe.call({
        method: "plantmaintenance.plantmaintenance.report.task_detail_action_flow.task_detail_action_flow.update_task_status",
        args: {
            task_ids: selected_tasks,
            status: status
        },
        callback: function(response) {
            if (response.message === "success") {
                frappe.msgprint(__("Task status updated successfully."));
                frappe.query_report.refresh();
            } else {
                frappe.msgprint(__("Failed to update task status."));
            }
        }
    });
}

File: plantmaintenance/plantmaintenance/report/task_detail_action_flow/task_detail_action_flow.py
# Copyright (c) 2025, LogicalDNA and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import date_diff, today

def execute(filters):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_data(filters):
    where_conditions = []
    filters_dict = {}

    if filters.get("from_date"):
        where_conditions.append("td.plan_start_date >= %(from_date)s")
        filters_dict["from_date"] = filters.get("from_date")

    if filters.get("to_date"):
        where_conditions.append("td.plan_start_date <= %(to_date)s")
        filters_dict["to_date"] = filters.get("to_date")

    if filters.get("task_detail"):
        where_conditions.append("td.name = %(task_detail)s")
        filters_dict["task_detail"] = filters.get("task_detail")

    if filters.get("type"):
        where_conditions.append("td.type = %(type)s")
        filters_dict["type"] = filters.get("type")

    if filters.get("work_center"):
        where_conditions.append("td.work_center = %(work_center)s")
        filters_dict["work_center"] = filters.get("work_center")

    if filters.get("equipment_group"):
        where_conditions.append("td.equipment_group = %(equipment_group)s")
        filters_dict["equipment_group"] = filters.get("equipment_group")

    if filters.get("status"):
        where_conditions.append("td.status = %(status)s")
        filters_dict["status"] = filters.get("status")

    where_clause = " AND ".join(where_conditions) if where_conditions else "1=1"
    query = f"""
        SELECT
            td.name AS task_detail,
            td.plan_start_date,
			td.old_tag_dcs,
            td.equipment_name,
            td.equipment_group,
			td.type,
            td.activity,
			td.parameter,
            td.assigned_to,
            td.send_for_approval_date,
            td.approved_date,
            td.completion_date,
            td.work_center,
            td.status,
            td.modified_by AS process_manager
           
        FROM
            `tabTask Detail` AS td
        LEFT JOIN
        `tabEquipment` AS eq ON td.old_tag_dcs = eq.name
        WHERE {where_clause}
    """
    raw_data = frappe.db.sql(query, filters_dict, as_dict=True)

    data = []

    for row in raw_data:
        assigned_to_name = frappe.db.get_value("User", {"name": row['assigned_to']}, "first_name") or row['assigned_to']

        data.append({
            'task_detail': row['task_detail'],
            'plan_start_date': row['plan_start_date'],
			'old_tag_dcs': row['old_tag_dcs'],
            'equipment_name': row['equipment_name'],
            'equipment_group': row['equipment_group'],
			'type': row['type'],
            'activity': row['activity'],
			'parameter': row['parameter'],
            'assigned_to': assigned_to_name,
            'work_center': row['work_center'],
            'status': row['status'],
        })

    return data



def get_columns():
    return [
		{
            "label": "",
            "fieldname": "select_task",
            "fieldtype": "Checkbox",
            "width": 50
        },
        {
            "label": "Task Id",
            "fieldname": "task_detail",
            "fieldtype": "Link",
            "options": "Task Detail",
            "width": 200
        },
        {
            "label": "Maintenance Type",
            "fieldname": "type",
            "fieldtype": "Select",
            "options": "Task Detail",
            "width": 200
        },
        {
            "label": "Status",
            "fieldname": "status",
            "fieldtype": "Select",
            "width": 200
        },
        {
            "label": "Plan Start Date",
            "fieldname": "plan_start_date",
            "fieldtype": "Date",
            "width": 150
        },
		{
            "label": "OLD TAG (DCS)",
            "fieldname": "old_tag_dcs",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Equipment Name",
            "fieldname": "equipment_name",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Equipment Group",
            "fieldname": "equipment_group",
            "fieldtype": "Link",
            "options": "Task Detail",
            "width": 150
        },
        {
            "label": "Work Center",
            "fieldname": "work_center",
            "fieldtype": "Link",
            "options": "Task Detail",
            "width": 200
        },
        {
            "label": "Activity",
            "fieldname": "activity",
            "fieldtype": "Link",
            "options": "Activity",
            "width": 200
        },
        {
            "label": "Parameter",
            "fieldname": "parameter",
            "fieldtype": "Link",
            "options": "Task Detail",
            "width": 200
        },
        {
            "label": "Assigned To",
            "fieldname": "assigned_to",
            "fieldtype": "Small Text",
            "width": 250
        },
       
    ]

@frappe.whitelist()
def update_task_status(task_ids, status):
    if isinstance(task_ids, str):
        task_ids = frappe.parse_json(task_ids)

    if not task_ids:
        return "No tasks selected"

    workflow_states = {
        "Approved": "Approved",
        "Rejected": "Rejected",
        "Completed": "Completed",
        "Cancelled": "Cancelled"
    }

    workflow_state = workflow_states.get(status)

    for task_id in task_ids:
        task_doc = frappe.get_doc("Task Detail", task_id)
        task_doc.status = status 
        if workflow_state:
            task_doc.workflow_state = workflow_state  
        task_doc.save(ignore_permissions=True)
        frappe.db.commit()

    return "success"
